core/lib/dal/src/basic_witness_input_producer_dal.rs

```diff
@@ -0,0 +1,168 @@
+    #[strum(serialize = "in_progress")]
+    InProgress,
+    /// The final (happy case) state we expect all jobs to end up. After the run is complete,
+    /// the job uploaded it's inputs, it lands in successful.
+    #[strum(serialize = "successful")]
+    Successful,
+    /// The job failed for reasons. It will be marked as such and the error persisted in DB.
+    /// If it failed less than MAX_ATTEMPTs, the job will be retried,
+    /// otherwise it will stay in this state as final state.
+    #[strum(serialize = "failed")]
+    Failed,
+}
+
+impl BasicWitnessInputProducerDal<'_, '_> {
+    pub async fn create_basic_witness_input_producer_job(
+        &mut self,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<()> {
+        sqlx::query!(
+            "INSERT INTO basic_witness_input_producer_jobs \
+                (l1_batch_number, status, created_at, updated_at) \
+            VALUES ($1, $2, now(), now())",
+            l1_batch_number.0 as i64,
+            BasicWitnessInputProducerJobStatus::Queued.to_string(),
+        )
+        .instrument("create_basic_witness_input_producer_job")
+        .report_latency()
+        .execute(self.storage.conn())
+        .await?;
+
+        Ok(())
+    }
+
+    pub async fn get_next_basic_witness_input_producer_job(
+        &mut self,
+    ) -> anyhow::Result<Option<L1BatchNumber>> {
+        let l1_batch_number = sqlx::query!(
+            "UPDATE basic_witness_input_producer_jobs \
+            SET status = $1, \
+                attempts = attempts + 1, \
+                updated_at = now(), \
+                processing_started_at = now() \
+            WHERE l1_batch_number = ( \
+                SELECT l1_batch_number \
+                FROM basic_witness_input_producer_jobs \
+                WHERE status = $2 OR \
+                    (status = $1 AND processing_started_at < now() - $4::interval) OR \
+                    (status = $3 AND attempts < $5) \
+                ORDER BY l1_batch_number ASC \
+                LIMIT 1 \
+                FOR UPDATE \
+                SKIP LOCKED \
+            )
+            RETURNING basic_witness_input_producer_jobs.l1_batch_number",
+            BasicWitnessInputProducerJobStatus::InProgress.to_string(),
+            BasicWitnessInputProducerJobStatus::Queued.to_string(),
+            BasicWitnessInputProducerJobStatus::Failed.to_string(),
+            &JOB_PROCESSING_TIMEOUT,
+            JOB_MAX_ATTEMPT,
+        )
+        .instrument("get_next_basic_witness_input_producer_job")
+        .report_latency()
+        .fetch_optional(self.storage.conn())
+        .await?
+        .map(|job| L1BatchNumber(job.l1_batch_number as u32));
+
+        Ok(l1_batch_number)
+    }
+
+    pub async fn mark_job_as_successful(
+        &mut self,
+        l1_batch_number: L1BatchNumber,
+        started_at: Instant,
+        object_path: String,
```

Comment:
Nit: You can use `&str` instead of `String` here; a string is not consumed by the method.